File: py/app/converters/openai_to_gemini.py
"""Convert OpenAI format to Gemini format."""

import json
from typing import Any

from py.app.models import gemini as gem
from py.app.models import openai as oai


def convert_request(req: oai.ChatCompletionRequest) -> tuple[gem.GenerateContentRequest, str]:
    """Convert OpenAI chat completion request to Gemini generate content request.

    Returns (request, model_name) tuple.
    """
    # Extract system instruction
    system_instruction = None
    contents: list[gem.Content] = []

    for msg in req.messages:
        if msg.role == "system":
            system_instruction = gem.SystemInstruction(
                parts=[gem.TextPart(text=msg.content or "")]
            )
        elif msg.role == "user":
            contents.append(
                gem.Content(role="user", parts=[gem.TextPart(text=msg.content or "")])
            )
        elif msg.role == "assistant":
            if msg.tool_calls:
                parts: list[gem.Part] = []
                if msg.content:
                    parts.append(gem.TextPart(text=msg.content))
                for tc in msg.tool_calls:
                    parts.append(
                        gem.FunctionCallPart(
                            functionCall={
                                "name": tc.function.name,
                                "args": json.loads(tc.function.arguments),
                            }
                        )
                    )
                contents.append(gem.Content(role="model", parts=parts))
            else:
                contents.append(
                    gem.Content(
                        role="model", parts=[gem.TextPart(text=msg.content or "")]
                    )
                )
        elif msg.role == "tool":
            # Tool result in Gemini format
            contents.append(
                gem.Content(
                    role="user",
                    parts=[
                        gem.FunctionResponsePart(
                            functionResponse={
                                "name": msg.name or "",
                                "response": json.loads(msg.content or "{}"),
                            }
                        )
                    ],
                )
            )

    # Convert tools
    tools = None
    if req.tools:
        function_declarations = []
        for tool in req.tools:
            function_declarations.append(
                gem.FunctionDeclaration(
                    name=tool.function.name,
                    description=tool.function.description,
                    parameters={
                        "type": tool.function.parameters.type
                        if tool.function.parameters
                        else "object",
                        "properties": tool.function.parameters.properties
                        if tool.function.parameters
                        else {},
                        "required": tool.function.parameters.required
                        if tool.function.parameters
                        else [],
                    }
                    if tool.function.parameters
                    else None,
                )
            )
        tools = [gem.ToolConfig(functionDeclarations=function_declarations)]

    # Generation config
    generation_config = gem.GenerationConfig(
        temperature=req.temperature,
        topP=req.top_p,
        maxOutputTokens=req.max_tokens,
        stopSequences=req.stop if isinstance(req.stop, list) else [req.stop] if req.stop else None,
    )

    return (
        gem.GenerateContentRequest(
            contents=contents,
            systemInstruction=system_instruction,
            generationConfig=generation_config,
            tools=tools,
        ),
        req.model,
    )


def convert_response(
    resp: gem.GenerateContentResponse, model: str
) -> oai.ChatCompletionResponse:
    """Convert Gemini generate content response to OpenAI chat completion response."""
    import time

    if not resp.candidates:
        return oai.ChatCompletionResponse(
            id=f"gemini-{int(time.time())}",
            created=int(time.time()),
            model=model,
            choices=[],
            usage=None,
        )

    candidate = resp.candidates[0]
    content = None
    tool_calls = None

    text_parts = []
    function_call_parts = []

    for part in candidate.content.parts:
        if isinstance(part, gem.TextPart):
            text_parts.append(part.text)
        elif isinstance(part, gem.FunctionCallPart):
            function_call_parts.append(part.functionCall)

    if text_parts:
        content = "".join(text_parts)

    if function_call_parts:
        tool_calls = []
        for i, fc in enumerate(function_call_parts):
            tool_calls.append(
                oai.ToolCall(
                    id=f"call_{i}",
                    type="function",
                    function=oai.FunctionCall(
                        name=fc.get("name", ""),
                        arguments=json.dumps(fc.get("args", {})),
                    ),
                )
            )

    # Map finish reason
    finish_reason_map = {
        "STOP": "stop",
        "MAX_TOKENS": "length",
        "SAFETY": "content_filter",
        "RECITATION": "content_filter",
    }
    finish_reason = finish_reason_map.get(candidate.finishReason or "STOP", "stop")
    if tool_calls:
        finish_reason = "tool_calls"

    # Usage
    usage = None
    if resp.usageMetadata:
        usage = oai.Usage(
            prompt_tokens=resp.usageMetadata.promptTokenCount or 0,
            completion_tokens=resp.usageMetadata.candidatesTokenCount or 0,
            total_tokens=resp.usageMetadata.totalTokenCount or 0,
        )

    return oai.ChatCompletionResponse(
        id=f"gemini-{int(time.time())}",
        created=int(time.time()),
        model=model,
        choices=[
            oai.Choice(
                index=0,
                message=oai.ChoiceMessage(
                    role="assistant",
                    content=content,
                    tool_calls=tool_calls,
                ),
                finish_reason=finish_reason,
            )
        ],
        usage=usage,
    )


def convert_stream_chunk(data: dict[str, Any], model: str) -> str | None:
    """Convert Gemini stream chunk to OpenAI stream chunk.

    Returns SSE formatted string or None if chunk should be skipped.
    """
    import time

    candidates = data.get("candidates", [])
    if not candidates:
        return None

    candidate = candidates[0]
    content = candidate.get("content", {})
    parts = content.get("parts", [])

    text_content = ""
    for part in parts:
        if "text" in part:
            text_content += part["text"]

    if text_content:
        chunk = oai.ChatCompletionChunk(
            id=f"gemini-{int(time.time())}",
            created=int(time.time()),
            model=model,
            choices=[
                oai.StreamChoice(
                    index=0,
                    delta=oai.DeltaMessage(content=text_content),
                    finish_reason=None,
                )
            ],
        )
        return f"data: {chunk.model_dump_json()}\n\n"

    finish_reason = candidate.get("finishReason")
    if finish_reason:
        finish_reason_map = {
            "STOP": "stop",
            "MAX_TOKENS": "length",
            "SAFETY": "content_filter",
        }
        chunk = oai.ChatCompletionChunk(
            id=f"gemini-{int(time.time())}",
            created=int(time.time()),
            model=model,
            choices=[
                oai.StreamChoice(
                    index=0,
                    delta=oai.DeltaMessage(),
                    finish_reason=finish_reason_map.get(finish_reason, "stop"),
                )
            ],
        )
        return f"data: {chunk.model_dump_json()}\n\ndata: [DONE]\n\n"

    return None

File: go/internal/converters/openai_to_gemini.go
package converters

import (
	"encoding/json"
	"time"

	"ai_gateway/internal/models"
)

// OpenAIToGeminiRequest converts an OpenAI request to Gemini format
func OpenAIToGeminiRequest(req *models.ChatCompletionRequest) (*models.GenerateContentRequest, error) {
	geminiReq := &models.GenerateContentRequest{}

	// Set generation config
	geminiReq.GenerationConfig = &models.GenerationConfig{}
	if req.Temperature != nil {
		geminiReq.GenerationConfig.Temperature = req.Temperature
	}
	if req.TopP != nil {
		geminiReq.GenerationConfig.TopP = req.TopP
	}
	if req.MaxTokens != nil {
		geminiReq.GenerationConfig.MaxOutputTokens = req.MaxTokens
	}

	// Convert stop sequences
	if req.Stop != nil {
		switch v := req.Stop.(type) {
		case string:
			geminiReq.GenerationConfig.StopSequences = []string{v}
		case []interface{}:
			for _, s := range v {
				if str, ok := s.(string); ok {
					geminiReq.GenerationConfig.StopSequences = append(geminiReq.GenerationConfig.StopSequences, str)
				}
			}
		}
	}

	// Convert messages
	var contents []models.GeminiContent
	for _, msg := range req.Messages {
		if msg.Role == "system" {
			// Extract system instruction
			content := getTextContent(msg.Content)
			geminiReq.SystemInstruction = &models.GeminiContent{
				Parts: []models.GeminiPart{{Text: content}},
			}
			continue
		}

		geminiContent := models.GeminiContent{}

		// Map role
		switch msg.Role {
		case "assistant":
			geminiContent.Role = "model"
		case "tool":
			// Tool response - add to previous model content or create new user content
			geminiContent.Role = "user"
			// Parse the content as function response
			var responseContent interface{}
			if str, ok := msg.Content.(string); ok {
				json.Unmarshal([]byte(str), &responseContent)
			} else {
				responseContent = msg.Content
			}
			geminiContent.Parts = []models.GeminiPart{{
				FunctionResponse: &models.FunctionResponse{
					Name:     msg.Name,
					Response: map[string]interface{}{"result": responseContent},
				},
			}}
			contents = append(contents, geminiContent)
			continue
		default:
			geminiContent.Role = "user"
		}

		// Handle tool calls from assistant
		if msg.ToolCalls != nil && len(msg.ToolCalls) > 0 {
			for _, tc := range msg.ToolCalls {
				var args map[string]interface{}
				json.Unmarshal([]byte(tc.Function.Arguments), &args)
				geminiContent.Parts = append(geminiContent.Parts, models.GeminiPart{
					FunctionCall: &models.GeminiFunctionCall{
						Name: tc.Function.Name,
						Args: args,
					},
				})
			}
		}

		// Handle regular content
		content := getTextContent(msg.Content)
		if content != "" {
			geminiContent.Parts = append(geminiContent.Parts, models.GeminiPart{Text: content})
		}

		if len(geminiContent.Parts) > 0 {
			contents = append(contents, geminiContent)
		}
	}
	geminiReq.Contents = contents

	// Convert tools
	if len(req.Tools) > 0 {
		var declarations []models.FunctionDeclaration
		for _, tool := range req.Tools {
			declarations = append(declarations, models.FunctionDeclaration{
				Name:        tool.Function.Name,
				Description: tool.Function.Description,
				Parameters:  tool.Function.Parameters,
			})
		}
		geminiReq.Tools = []models.GeminiTool{{
			FunctionDeclarations: declarations,
		}}
	}

	return geminiReq, nil
}

// GeminiToOpenAIResponse converts a Gemini response to OpenAI format
func GeminiToOpenAIResponse(resp map[string]interface{}, model string) (*models.ChatCompletionResponse, error) {
	openaiResp := &models.ChatCompletionResponse{
		ID:      generateID(),
		Object:  "chat.completion",
		Created: time.Now().Unix(),
		Model:   model,
	}

	candidates, ok := resp["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return openaiResp, nil
	}

	candidate := candidates[0].(map[string]interface{})
	content := candidate["content"].(map[string]interface{})
	parts := content["parts"].([]interface{})

	var message models.ChatMessage
	message.Role = "assistant"

	var textContent string
	var toolCalls []models.ToolCall
	toolCallIndex := 0

	for _, part := range parts {
		partMap := part.(map[string]interface{})
		if text, ok := partMap["text"].(string); ok {
			textContent += text
		}
		if fc, ok := partMap["functionCall"].(map[string]interface{}); ok {
			args, _ := json.Marshal(fc["args"])
			toolCalls = append(toolCalls, models.ToolCall{
				ID:   generateToolCallID(toolCallIndex),
				Type: "function",
				Function: models.FunctionCall{
					Name:      getString(fc, "name"),
					Arguments: string(args),
				},
			})
			toolCallIndex++
		}
	}

	if textContent != "" {
		message.Content = textContent
	}
	if len(toolCalls) > 0 {
		message.ToolCalls = toolCalls
	}

	// Convert finish reason
	var finishReason string
	if fr, ok := candidate["finishReason"].(string); ok {
		switch fr {
		case "STOP":
			finishReason = "stop"
		case "MAX_TOKENS":
			finishReason = "length"
		default:
			if len(toolCalls) > 0 {
				finishReason = "tool_calls"
			} else {
				finishReason = "stop"
			}
		}
	}

	openaiResp.Choices = []models.Choice{{
		Index:        0,
		Message:      &message,
		FinishReason: &finishReason,
	}}

	// Convert usage
	if usage, ok := resp["usageMetadata"].(map[string]interface{}); ok {
		promptTokens := getInt(usage, "promptTokenCount")
		completionTokens := getInt(usage, "candidatesTokenCount")
		openaiResp.Usage = &models.Usage{
			PromptTokens:     promptTokens,
			CompletionTokens: completionTokens,
			TotalTokens:      promptTokens + completionTokens,
		}
	}

	return openaiResp, nil
}

// GeminiStreamToOpenAIStream converts a Gemini stream event to OpenAI format
func GeminiStreamToOpenAIStream(data map[string]interface{}, model string, id string) ([]byte, error) {
	candidates, ok := data["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return nil, nil
	}

	candidate := candidates[0].(map[string]interface{})
	content, ok := candidate["content"].(map[string]interface{})
	if !ok {
		return nil, nil
	}

	parts, ok := content["parts"].([]interface{})
	if !ok || len(parts) == 0 {
		return nil, nil
	}

	chunk := models.ChatCompletionChunk{
		ID:      id,
		Object:  "chat.completion.chunk",
		Created: time.Now().Unix(),
		Model:   model,
	}

	part := parts[0].(map[string]interface{})
	if text, ok := part["text"].(string); ok {
		chunk.Choices = []models.Choice{{
			Index: 0,
			Delta: &models.ChatMessage{Content: text},
		}}
	} else if fc, ok := part["functionCall"].(map[string]interface{}); ok {
		args, _ := json.Marshal(fc["args"])
		chunk.Choices = []models.Choice{{
			Index: 0,
			Delta: &models.ChatMessage{
				ToolCalls: []models.ToolCall{{
					ID:   generateToolCallID(0),
					Type: "function",
					Function: models.FunctionCall{
						Name:      getString(fc, "name"),
						Arguments: string(args),
					},
				}},
			},
		}}
	}

	// Check for finish reason
	if fr, ok := candidate["finishReason"].(string); ok {
		var finishReason string
		switch fr {
		case "STOP":
			finishReason = "stop"
		case "MAX_TOKENS":
			finishReason = "length"
		default:
			finishReason = "stop"
		}
		if len(chunk.Choices) > 0 {
			chunk.Choices[0].FinishReason = &finishReason
		}
	}

	return json.Marshal(chunk)
}

func generateToolCallID(index int) string {
	return "call_" + time.Now().Format("20060102150405") + "_" + string(rune('a'+index))
}
